### PR TITLE
Add Explain button to SQL log panel

### PR DESCRIPTION
**Pull request by chinpei215**
After cakephp/cakephp#9421 merged, DebugKit will show you Explain buttons in SQL log panel. I confirmed working with Vivaldi 1.3, IE 11, Edge 25, Chrome 53 and FF 48, on WIndows.

### MySQL
explain-mysql

### PostgreSQL
explain-postgres

### SQLite
explain-sqlite

### CODE REVIEW

File: src/Controller/ToolbarController.php [hunk 1]
```diff
@@ -71,4 +74,57 @@ public function clearCache()
...
+        $data = $this->request->data('data');
+        $hash = $this->request->data('hash');
+
+        if (Security::hash($data, null, true) !== $hash) {
+            throw new BadRequestException('Invalid hash');
+        }
+
+        $data = json_decode($data, true);
+        if (!$data) {
+            throw new BadRequestException('Invalid json');
```
Comment (dereuromark) [inline src/Controller/ToolbarController.php]: JSON

**Comment by dereuromark** [comment]
Nice :)

File: src/Template/Element/sql_log_panel.ctp [hunk 2]
```diff
@@ -65,14 +71,38 @@ SqlFormatter::$pre_attributes = 'style="color: #222; background-color: transpare
...
+                        $data = json_encode([
+                            'connection' => $logger->name(),
+                            'query' => $query['queryString'],
+                            'params' => $query['params']
+                        ]);
+                        echo $this->Form->create();
+                        echo $this->Form->hidden('data', ['value' => $data]);
+                        echo $this->Form->hidden('hash', ['value' => Security::hash($data, null, true)]);
+                        echo $this->Form->button('Explain', ['type' => 'button', 'class' => 'btn-primary sql-explain-link']);
+                        echo $this->Form->end();
```
Comment (markstory) [inline src/Template/Element/sql_log_panel.ctp]: Is this something we could use postLink() or postButton() for?

File: src/Template/Element/sql_log_panel.ctp [hunk 3]
```diff
@@ -84,3 +114,57 @@ SqlFormatter::$pre_attributes = 'style="color: #222; background-color: transpare
...
+<script>
+$(document).ready(function() {
+    var baseUrl = '<?= $this->Url->build([
+        'plugin' => 'DebugKit',
+        'controller' => 'Toolbar',
+        'action' => 'sqlExplain'
+    ]); ?>';
+
+    function createTable(result) {
+        var $table = $('<table>');
```
Comment (markstory) [inline src/Template/Element/sql_log_panel.ctp]: I don't think the `$` prefixes are necessary here.

**Comment by markstory** [comment]
Should we update the composer.json to require the new version of cakephp/cakephp? That might remove the need for `method_exists` checks.

**Comment by ravage84** [comment]
@markstory might be an option, but I wouldn't do that in a patch release then...

**Comment by markstory** [comment]
@ravage84 I was going to make the explain button go into a minor release of DebugKit. Its a pretty big new feature.

**Comment by chinpei215** [comment]
Thank you for your comments.

> Should we update the composer.json to require the new version of cakephp/cakephp? That might remove the need for method_exists checks.

I am fine to target next minor release. Then we can remove `isset($query->queryString)` check from DebugLog. However, we would still need to call `method_exists`, as the canExplain method is not contained in ConnectionInterface.

File: src/Template/Element/sql_log_panel.ctp (hunk 3 above)
Comment (chinpei215) [inline src/Template/Element/sql_log_panel.ctp]: Ah sorry. It's my personal habit 😛 When I assign an instance of jQuery object assign to a variable, I always prefix `$` to the name. I will remove all of them.

**Pull request closed without merging**